File: priv/src/exdl_rtf.c
/*
 *  Copyright (c) 2008 David Koch
 *
 *  See the file "license.terms" for information on usage and redistribution
 *  of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 *     $Id: exdl_rtf.c,v 1.0 2008/12/29 14:45:38 kochise Exp $
 */

#include "exdl.h"

#ifdef COMPILE_RTF

#include <string.h>
#include <stdlib.h>

EXDL_API(ertf_linked_Version)
{
  /* const SDL_version* RTF_Linked_Version(void) */

  /* exp : Get linked library's version. */
  /* inp :  */
  /* ret :  */

  CATCH_RESULT(const SDL_version*)

    RTF_Linked_Version
    (
    );

  THROW_POINTER()
}

EXDL_API(ertf_createContext)
{
  /* RTF_Context* RTF_CreateContext(RTF_FontEngine* fontEngine) */

  /* exp : Create an RTF display context, with the given font engine. Once a context */
  /*       is created, it can be used to load and display text in Microsoft RTF format. */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(RTF_FontEngine, fontEngine[])
  STACK_FMT_END()

  CATCH_RESULT(RTF_Context*)

    RTF_CreateContext
    ( STACK_GET_ADDR(fontEngine)
    );

  THROW_POINTER()
}

EXDL_API(ertf_load)
{
  /* int RTF_Load(RTF_Context* ctx, const char* file) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(RTF_Context*, ctx)
    STACK_FMT_DATA(char,         file[])
  STACK_FMT_END()

  CATCH_RESULT(int)

    RTF_Load
    ( STACK_GET_DATA(ctx)
    , STACK_GET_ADDR(file)
    );

  THROW_RESULT(int)
}

EXDL_API(ertf_load_RW)
{
  /* int RTF_Load_RW(RTF_Context* ctx, SDL_RWops* src, int freesrc) */

  /* exp : Set the text of an RTF context. */
  /*     > Use RTF_GetError() to get a text message corresponding to the error. */
  /* inp :  */
  /* ret : 0 if it succeeds or -1 if it fails. */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(RTF_Context*, ctx)
    STACK_FMT_LPTR(SDL_RWops*,   src)
    STACK_FMT_DATA(int,          freesrc)
  STACK_FMT_END()

  CATCH_RESULT(int)

    RTF_Load_RW
    ( STACK_GET_DATA(ctx)
    , STACK_GET_DATA(src)
    , STACK_GET_DATA(freesrc)
    );

  THROW_RESULT(int)
}

EXDL_API(ertf_getTitle)
{
  /* const char* RTF_GetTitle(RTF_Context* ctx) */

  /* exp : Get the title of an RTF document */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(RTF_Context*, ctx)
  STACK_FMT_END()

  CATCH_RESULT(const char*)

    RTF_GetTitle
    ( STACK_GET_DATA(ctx)
    );

  THROW_STRING()
}

EXDL_API(ertf_getSubject)
{
  /* const char* RTF_GetSubject(RTF_Context* ctx) */

  /* exp : Get the subject of an RTF document */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(RTF_Context*, ctx)
  STACK_FMT_END()

  CATCH_RESULT(const char*)

    RTF_GetSubject
    ( STACK_GET_DATA(ctx)
    );

  THROW_STRING()
}

EXDL_API(ertf_getAuthor)
{
  /* const char* RTF_GetAuthor(RTF_Context* ctx) */

  /* exp : Get the author of an RTF document */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(RTF_Context*, ctx)
  STACK_FMT_END()

  CATCH_RESULT(const char*)

    RTF_GetAuthor
    ( STACK_GET_DATA(ctx)
    );

  THROW_STRING()
}

EXDL_API(ertf_getHeight)
{
  /* int RTF_GetHeight(RTF_Context* ctx, int width) */

  /* exp : Get the height of an RTF render area given a certain width. The text is */
  /*       automatically reflowed to this new width, and should match the width of */
  /*       the clipping rectangle used for rendering later. */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(RTF_Context*, ctx)
    STACK_FMT_DATA(int,          width)
  STACK_FMT_END()

  CATCH_RESULT(int)

    RTF_GetHeight
    ( STACK_GET_DATA(ctx)
    , STACK_GET_DATA(width)
    );

  THROW_RESULT(int)
}

EXDL_API(ertf_render)
{
  /* void RTF_Render(RTF_Context* ctx, SDL_Surface* surface, SDL_Rect* rect, int yOffset) */

  /* exp : Render the RTF document to a rectangle of a surface. The text is reflowed */
  /*       to match the width of the rectangle. The rendering is offset up (and */
  /*       clipped) by yOffset pixels. */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(RTF_Context*, ctx)
    STACK_FMT_LPTR(SDL_Surface*, surface)
    STACK_FMT_LPTR(SDL_Rect*,    rect)
    STACK_FMT_DATA(int,          yOffset)
  STACK_FMT_END()

    RTF_Render
    ( STACK_GET_DATA(ctx)
    , STACK_GET_DATA(surface)
    , STACK_GET_DATA(rect)
    , STACK_GET_DATA(yOffset)
    );
}

EXDL_API(ertf_freeContext)
{
  /* void RTF_FreeContext(RTF_Context* ctx) */

  /* exp : Free an RTF display context */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(RTF_Context, ctx[])
  STACK_FMT_END()

    RTF_FreeContext
    ( STACK_GET_ADDR(ctx)
    );
}

#endif // COMPILE_RTF

File: src/exdl_rtf.erl
%%
%%  Copyright (c) 2008 David Koch
%%
%%  See the file "license.terms" for information on usage and redistribution
%%  of this file, and for a DISCLAIMER OF ALL WARRANTIES.
%% 
%%     $Id: exdl_rtf.erl,v 1.0 2008/12/30 07:46:20 kochise Exp $
%%

-module(exdl_rtf).

-include("exdl.hrl").

-ifdef(COMPILE_RTF).

%% 1- EXPORT

%-compile(export_all).

-export( [ linkedVersion/0
         , createContext/1
         , load/2
         , loadRW/3
         , getTitle/1
         , getSubject/1
         , getAuthor/1
         , getHeight/2
         , render/4
         , freeContext/1
         ]
       ).

%% 2- INCLUDE

-include("exdl_rtf.hrl").

%% 3- IMPORT

-import( exdl_driver
       , [ call/2
         , cast/2
         ]
       ).

%% 4- FUNCTIONS

%% Func:       linkedVersion
%% Args:       none
%% Returns:    Exdl
%% C-API func: const SDL_version* RTF_Linked_Version(void)

linkedVersion()
 -> case
      call
      ( ?RTF_Linked_Version
      , [
        ]
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?_PTR
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       createContext
%% Args:       FontEngine
%% Returns:    Exdl
%% C-API func: RTF_Context* RTF_CreateContext(RTF_FontEngine* fontEngine)

createContext(FontEngine)
 -> case
      call
      ( ?RTF_CreateContext
      , <<FontEngine:?_PTR
        >>
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?_PTR
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       load
%% Args:       Ctx, File
%% Returns:    Exdl
%% C-API func: int RTF_Load(RTF_Context* ctx, const char* file)

load(Ctx, File)
 -> case
      call
      ( ?RTF_Load
      , <<Ctx:?_PTR
        , File:?_PTR
        >>
      )
    of
      <<Exdl:?Int
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       loadRW
%% Args:       Ctx, Src, Freesrc
%% Returns:    Exdl
%% C-API func: int RTF_Load_RW(RTF_Context* ctx, SDL_RWops* src, int freesrc)

loadRW(Ctx, Src, Freesrc)
 -> case
      call
      ( ?RTF_Load_RW
      , <<Ctx:?_PTR
        , Src:?_PTR
        , Freesrc:?Int
        >>
      )
    of
      <<Exdl:?Int
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       getTitle
%% Args:       Ctx
%% Returns:    Exdl
%% C-API func: const char* RTF_GetTitle(RTF_Context* ctx)

getTitle(Ctx)
 -> case
      call
      ( ?RTF_GetTitle
      , <<Ctx:?_PTR
        >>
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?String
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       getSubject
%% Args:       Ctx
%% Returns:    Exdl
%% C-API func: const char* RTF_GetSubject(RTF_Context* ctx)

getSubject(Ctx)
 -> case
      call
      ( ?RTF_GetSubject
      , <<Ctx:?_PTR
        >>
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?String
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       getAuthor
%% Args:       Ctx
%% Returns:    Exdl
%% C-API func: const char* RTF_GetAuthor(RTF_Context* ctx)

getAuthor(Ctx)
 -> case
      call
      ( ?RTF_GetAuthor
      , <<Ctx:?_PTR
        >>
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?String
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       getHeight
%% Args:       Ctx, Width
%% Returns:    Exdl
%% C-API func: int RTF_GetHeight(RTF_Context* ctx, int width)

getHeight(Ctx, Width)
 -> case
      call
      ( ?RTF_GetHeight
      , <<Ctx:?_PTR
        , Width:?Int
        >>
      )
    of
      <<Exdl:?Int
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       render
%% Args:       Ctx, Surface, Rect, YOffset
%% Returns:    ok
%% C-API func: void RTF_Render(RTF_Context* ctx, SDL_Surface* surface, SDL_Rect* rect, int yOffset)

render(Ctx, Surface, Rect, YOffset)
 -> cast
    ( ?RTF_Render
    , <<Ctx:?_PTR
      , Surface:?_PTR
      , Rect:?_PTR
      , YOffset:?Int
      >>
    )
.

%% Func:       freeContext
%% Args:       Ctx
%% Returns:    ok
%% C-API func: void RTF_FreeContext(RTF_Context* ctx)

freeContext(Ctx)
 -> cast
    ( ?RTF_FreeContext
    , <<Ctx:?_PTR
      >>
    )
.

-endif. % COMPILE_RTF
